### ISSUE TITLE
Recording failed to start (time-out)

### ISSUE THREAD

**Issue by codal-mharkhani** [jitsi/jibri#293]
Hi Team,

I am running Jitsi-meet and Jibri in two separate VMs.
When clicking on Start Recording it says "preparing to record the meeting" and after a while "Recording failed to start".

`2020-05-13 09:30:04.419 SEVERE: [117] org.jitsi.jibri.selenium.pageobjects.CallPage.visit() Timed out waiting for call page to load
2020-05-13 09:30:04.419 INFO: [117] org.jitsi.jibri.selenium.JibriSelenium.onSeleniumStateChange() Transitioning from state Starting up to Error: SESSION Failed to join call
2020-05-13 09:30:04.420 INFO: [117] org.jitsi.jibri.service.impl.FileRecordingJibriService.onServiceStateChange() File recording service transitioning from state Starting up to Error: SESSION Failed to join call
2020-05-13 09:30:04.420 INFO: [117] org.jitsi.jibri.api.xmpp.XmppApi.invoke() Current service had an error, sending error iq <iq to='jibribrewery@internal.auth.dev-meet.hiosocial.com/focus' id='R69Qv-140' type='set'><jibri xmlns='http://jitsi.org/protocol/jibri' status='off' failure_reason='error'/></iq>
2020-05-13 09:30:04.420 FINE: [117] org.jitsi.jibri.statsd.JibriStatsDClient.incrementCounter() Incrementing statsd counter: stop:recording
2020-05-13 09:30:04.423 INFO: [117] org.jitsi.jibri.JibriManager.stopService() Stopping the current service
2020-05-13 09:30:04.423 INFO: [117] org.jitsi.jibri.service.impl.FileRecordingJibriService.stop() Stopping capturer
2020-05-13 09:30:04.423 INFO: [117] org.jitsi.jibri.util.JibriSubprocess.ffmpeg.stop() Stopping ffmpeg process
2020-05-13 09:30:04.423 INFO: [117] org.jitsi.jibri.util.JibriSubprocess.ffmpeg.stop() ffmpeg exited with value null
2020-05-13 09:30:04.423 INFO: [117] org.jitsi.jibri.service.impl.FileRecordingJibriService.stop() Quitting selenium
2020-05-13 09:30:04.430 INFO: [117] org.jitsi.jibri.service.impl.FileRecordingJibriService.stop() Participants in this recording: []
2020-05-13 09:30:04.449 INFO: [117] org.jitsi.jibri.selenium.JibriSelenium.leaveCallAndQuitBrowser() Got 10 log entries for type browser
2020-05-13 09:30:04.469 INFO: [117] org.jitsi.jibri.selenium.JibriSelenium.leaveCallAndQuitBrowser() Got 1055 log entries for type driver
2020-05-13 09:30:04.582 INFO: [117] org.jitsi.jibri.selenium.JibriSelenium.leaveCallAndQuitBrowser() Got 0 log entries for type client
2020-05-13 09:30:04.582 INFO: [117] org.jitsi.jibri.selenium.JibriSelenium.leaveCallAndQuitBrowser() Leaving web call
2020-05-13 09:30:04.608 INFO: [117] org.jitsi.jibri.selenium.JibriSelenium.leaveCallAndQuitBrowser() Quitting chrome driver
2020-05-13 09:30:04.677 INFO: [117] org.jitsi.jibri.selenium.JibriSelenium.leaveCallAndQuitBrowser() Chrome driver quit
2020-05-13 09:30:04.677 INFO: [117] org.jitsi.jibri.service.impl.FileRecordingJibriService.stop() Finalizing the recording
2020-05-13 09:30:04.678 SEVERE: [117] org.jitsi.jibri.service.impl.FileRecordingJibriService.finalize() Failed to run finalize script: java.io.IOException: Cannot run program "/path/to/finalize_recording.sh": error=2, No such file or directory with stack: 
java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)`

@damencho

**Comment by octoquad** [comment]
I've spent the last 24 hours getting this working and finally have it working perfectly. Some things to watch out for:

1. Once you have configured Jibri, restart prosody, jicofo and jitsi-video-bridge services.
2. Start / restart Jibri. Give it 90 seconds to register with Jitsi.
3. If the above doesn't fix your problem, make sure the Jibri configuration and Prosody user accounts are registered and configured correctly and repeat 1-2.

**Comment by codal-mharkhani** [comment]
@octoquad please can you verify my configure
[sip-communicator.properties.txt](https://github.com/jitsi/jibri/files/4621880/sip-communicator.properties.txt)
[dev-meet.hiosocial.com.cfg.lua.txt](https://github.com/jitsi/jibri/files/4621881/dev-meet.hiosocial.com.cfg.lua.txt)
[config.json.txt](https://github.com/jitsi/jibri/files/4621882/config.json.txt)

**Comment by octoquad** [comment]
@codal-mharkhani, your configs look correct. 

The documentation may be a bit outdated, this prosodyctl command:

`prosodyctl register jibri auth.yourdomain.com jibriauthpass`

Should actually be or the following should also be additionally run:

`prosodyctl register jibri internal.auth.yourdomain.com jibriauthpass`

As soon as I did that and restart Jibri it was registering correctly. However, I did pick up that the username I used in the jibri config was `jitsi` instead of `recorder` for the `call_login` object. So perhaps the above extra `prosdyctl` step is not required. Give it shot though and let me know.

**Comment by codal-mharkhani** [comment]
@octoquad thanks for the prompt reply.

Still facing the same issue, I'm attaching my prosody error.

[prosody.err.txt](https://github.com/jitsi/jibri/files/4624079/prosody.err.txt)

**Comment by octoquad** [comment]
Interesting, I have separate lua configuration for jibri called jibri-domain.com.cfg.lua which has:

```lua
Component "internal.auth.domain.com" "muc"
    modules_enabled = {
        "ping";
    }
    storage = "null"
    muc_room_cache_size = 1000

VirtualHost "recorder.domain.com"
    modules_enabled = {
	"ping";
    }
    authentication = "internal_plain"
```

And in the domain.cfg.lua:
```lua
...
Component "internal.auth.domain.com" "muc"
    storage = "memory"
    modules_enabled = {
      "ping";
    }
    admins = { "focus@auth.domain.com", "jvb@auth.domain.com" }
    muc_room_locking = false
    muc_room_default_public_jids = true
...
```

The documentation does say add a new component for Jibri, not re-use the existing one that is present for Jitsi.

**Comment by paddatrapper** [comment]
> I'm getting Error: No data storage active

I get the same with the config @octoquad provided. This can be fixed by setting `storage = "memory"`, but this doesn't fix the original issue of recordings and live streams failing to start.

I think that splitting the config in `/etc/prosody/conf.d` doesn't matter, as the last included Component with the name will be used. They aren't merged together or held separately.

**Comment by octoquad** [comment]
> I think that splitting the config in `/etc/prosody/conf.d` doesn't matter, as the last included Component with the name will be used. They aren't merged together or held separately.

Correct, this is just how most configurations are done to avoid having to diff against the package maintainers version for a main configuration file (although in this case it's generated based off the domain, so not relevant). For me, I can just focus on the component I want to work on which has a dedicated file.

Regarding "They aren't merged together or held separately.", the way I see it, the Component is read then assigned to the VirtualHost, but I could be wrong. 

Perhaps a dev can clarify how two Components with the same identifier work.

I also think [this section](https://github.com/jitsi/jibri#prosody) of the documentation might be confusing for some and can probably be clarified. For instance it says, add, not replace or reuse the MUC component from Jitsi for Jibri.

**Comment by octoquad** [comment]
By the way is Jibri running on the same host or different host? What OS and version are both Jitsi and Jibri running on. 

My setup is two Ubuntu 20.04 installations hosting Jitsi and Jibri (JRE 8 not JRE 11 for Jibri) separately. Ubuntu 20.04 Jitsi with 16.04 Jibri didn't work for me.

**Comment by paddatrapper** [comment]
I'm running Jibri on 16.04 and Jitsi on Debian 10 (Buster). Java OpenJDK 11 for Jitsi and OpenJDK 8 for Jibri

**Comment by octoquad** [comment]
So one problem I did have around this was SSL. I initially set up a 16.04 and 20.04 server in parallel, since 16.04 will be EOL in less than a year. When I ran into the original posters issue, I started looking at the server certificate just to make sure everything was OK. I ran `openssl s_client -connect domain.com:443` on the 16.04 server to the 20.04 server and got an error (handshake failure). On the 20.04 install for Jibri though, I got a successful SSL connection (no handshake failure) although no certificate information was provided. 

I also never really dived further on how to fix the SSL connection between 16.04 and 20.04, so I moved on with getting Jibri working on 20.04 instead. 

I also created a self-signed certificate for the recorder virtual host with `prosodyctl cert generate recorder.domain.com`. At one point I thought this might be an issue as well, but was not documented in the README. In the end, I think the SSL versions between hosts was probably the reason for the failure rather than a cert for the recorder virtualhost, which I will probably remove at some point.

So having said that, those would be the last two things that would be different between my working installation and yours. I'm not sure if I can be of any more help if those last two things I did, do not work for you.

I would also consider carefully reviewing the Jibri config and re-register the users on `prosodyctl` one last time and restarting all three services afterwards then firing up Jibri.

Good luck though! :)

**Comment by codal-mharkhani** [comment]
@octoquad Thank you for the prompt reply.

I follow your instruction but still have the same error, I check all the log and we are getting the error on  log.0.text and browser.o.txt

I think there has no problem with jitsi server because authentication is working fine. and the log URL I can open using a browser, as well on jibri server we are getting all the log file, is there any version conflict.

**Comment by matthiasg** [comment]
Where can I find documentation on these variables like

```
    muc_room_cache_size = 1000
    muc_room_locking = false
    muc_room_default_public_jids = true
   storage = "memory" / "null"
``

For example, what is cache_size doing exactly etc. Also when I use storage = "memory" I get errors in the log about storage_memory not being found.

and why is there a domain.cfg.lua in addition and how do they work together ?

**Comment by paddatrapper** [comment]
@matthiasg have a look at `/var/log/prosody/prosody.log`. There I found everything except jibri getting authenticated, but jibri never got further than `New BOSH session, assigned it sid`

Prosody is an XMPP server, its [documentation](https://prosody.im/doc/configure) should have explanations for those settings.

Prosody loads all additional configuration in `/etc/prosody/conf.d/*.cfg.lua`. They are merged into `/etc/prosody/prosody.cfg.lua` and ready by Prosody when it starts. The Jitsi deb package installs `/etc/prosody/conf.d/domain.com.cfg.lua`. The docs recommend using a different file for your specific configuration so that it does not cause issues if Jitsi upstream changes their configuration file.

**Comment by paddatrapper** [comment]
@octoquad that is interesting. I'll try replace 16.04 with Debian 10 and see how that goes.

**Comment by paddatrapper** [comment]
I managed to get past the timeout. The issue was that the recorder wasn't authenticating against `internal.auth.domain.org`, rather `recorder.domain.org`, so you need to register the recorder with that:

```
prosodyctl register recorder recorder.domain.org $password
```

Which I see is in the docs, but for some reason I hadn't run it. @codal-mharkhani that may help you too?

**Comment by codal-mharkhani** [comment]
> I've spent the last 8 hours getting this working and finally have it working perfectly. Some things to watch out for:
> 
> 1. Once you have configured Jibri, restart prosody, jicofo and jitsi-video-bridge services.
> 2. Start / restart Jibri. Give it 90 seconds to register with Jitsi.
> 3. If the above doesn't fix your problem, make sure the Jibri configuration and Prosody user accounts are registered and configured correctly and repeat 1-2.

@octoquad, I tried above step again and its working perfect, but after ended the calling user is not able to download the call

**Comment by codal-mharkhani** [comment]
@octoquad @matthiasg @paddatrapper 

is there any way to the user can download the recorded video without DropBox on call end action

**Comment by octoquad** [comment]
@codal-mharkhani, as far as I know, there is no way to download the recording from the Jitsi UI. There is a `finalize_recording.sh` script in the Jibri configuration that needs to be created. There is also a [jitsi-upload-integrations](https://github.com/jitsi/jitsi-upload-integrations) package that provides the Dropbox integration.

In my case, the recordings are copied from the `/tmp/recordings` directory on to a NFS mount point using a simple bash script:

```bash
#!/bin/bash

RECORDING=$(find $1 -name *.mp4)
DEST_DIR='/srv/recordings/'

mv "$RECORDING" "$DEST_DIR"
```

The above bash script is a minor adaptation from https://community.jitsi.org/t/jibri-showing-link-after-recording/53994

**Comment by octoquad** [comment]
@matthiasg, @paddatrapper, @codal-mharkhani. Since this issue was more of a configuration issue rather than a bug, do you have any feedback regarding the documentation to make it clearer? If not, I think this issue can possibly be closed.

**Comment by paddatrapper** [comment]
If possible, the error should be clearer in the logs. A timeout can have a load of causes, but a incorrect username/password is easier to track down.

It would also be great if browser.0.log was written to as it happened and not all at once at the end of the stream so it could be `tail -f`'ed instead of `cat` and search through slowly after the run

**Comment by bilogic** [comment]
Hi,

1. I'm trying to get the official docker to work using `docker compose -f jitsi-unzip/docker-compose.yml -f jitsi-unzip/jibri.yml up -d`
2. Earlier, my jibri account was unable to login until I used `register jibri internal.auth.meet.jitsi jibripassword`
3. Now, I see all my accounts logging in, except for `recorder`
    ```
    2025-02-18 02:50:59 meet.jitsi:muc_breakout_rooms                                info   Hook to muc events on muc.meet.jitsi
    2025-02-18 02:50:59 metadata.meet.jitsi:room_metadata_component                  info   Hook to muc events on muc.meet.jitsi
    2025-02-18 02:50:59 avmoderation.meet.jitsi:av_moderation_component              info   Hook to muc events on muc.meet.jitsi
    2025-02-18 02:50:59 speakerstats.meet.jitsi:speakerstats_component               info   Conference component loaded muc.meet.jitsi
    2025-02-18 02:50:59 speakerstats.meet.jitsi:speakerstats_component               info   Hook to muc events on muc.meet.jitsi
    2025-02-18 02:50:59 speakerstats.meet.jitsi:speakerstats_component               info   Main muc service table: 0x616c1e388920
    2025-02-18 02:51:00 c2s616c1e57eae0                                              info   Client connected
    2025-02-18 02:51:00 c2s616c1e57eae0                                              info   Stream encrypted (TLSv1.3 with TLS_AES_256_GCM_SHA384)
    2025-02-18 02:51:01 c2s616c1e58e030                                              info   Client connected
    2025-02-18 02:51:01 c2s616c1e57eae0                                              info   Authenticated as focus@auth.meet.jitsi
    2025-02-18 02:51:01 auth.meet.jitsi:limits_exception                             info   Setting stanza size limits for focus@auth.meet.jitsi to 10485760
    2025-02-18 02:51:01 c2s616c1e58e030                                              info   Stream encrypted (TLSv1.3 with TLS_AES_256_GCM_SHA384)
    2025-02-18 02:51:01 c2s616c1e58e030                                              info   Authenticated as jvb@auth.meet.jitsi
    2025-02-18 02:51:01 auth.meet.jitsi:limits_exception                             info   Setting stanza size limits for jvb@auth.meet.jitsi to 10485760
    2025-02-18 02:51:01 c2s616c1e641650                                              info   Client connected
    2025-02-18 02:51:01 c2s616c1e641650                                              info   Authenticated as 6d723108-d864-4e15-8171-ce49215e4aa3@guest.meet.jitsi
    2025-02-18 02:51:01 c2s616c1e641650                                              info   Client disconnected: connection closed
    2025-02-18 02:51:02 c2s616c1e65cb50                                              info   Client connected
    2025-02-18 02:51:02 c2s616c1e65cb50                                              info   Stream encrypted (TLSv1.3 with TLS_AES_256_GCM_SHA384)
    2025-02-18 02:51:02 c2s616c1e65cb50                                              info   Authenticated as jibri@auth.meet.jitsi
    2025-02-18 02:51:28 c2s616c1e7f24c0                                              info   Client connected
    2025-02-18 02:51:28 c2s616c1e7f24c0                                              info   Authenticated as 159369aa-308e-4408-ae69-95d0127a9c30@guest.meet.jitsi
    2025-02-18 02:51:36 c2s616c1e0d5070                                              info   Client connected
    2025-02-18 02:51:36 c2s616c1e0d5070                                              info   Authenticated as bilogic@meet.jitsi
    2025-02-18 02:51:37 c2s616c1e0d5070                                              info   Client disconnected: connection closed
    ```
4. I tried recording in my meeting, but it seems to allow only local recording and the red box CANNOT be toggled off
    ![Image](https://github.com/user-attachments/assets/ba691a72-0f88-404d-81df-e20be14a93f9)
5. When will `recorder` login, and how is jibri recording supposed to work?

Thanks!

**Comment by bilogic** [comment]
I managed to nail it, bad `.env` file, thanks!
